lib/oauth2-client.js

```diff
@@ -0,0 +1,123 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as storage from './storage.js';
+import * as bedrock from 'bedrock';
+import {httpClient} from '@digitalbazaar/http-client';
+import {LruCache} from '@digitalbazaar/lru-memoize';
+import logger from './logger.js';
+const {util: {BedrockError}} = bedrock;
+
+export class OAuth2Client {
+  /**
+   * Constructs an OAuth2Client instance to return an access token. It first
+   * checks the database to see if an access token already exists for the client
+   * id; if not, it creates an access token then stores it in the database.
+   *
+   * @param {object} options - Options to use.
+   * @param {string} options.config - The "oauth2-client-grant" properties.
+   * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+   *
+   * @returns {OAuth2Client} The `OAuth2Client` instance.
+   */
+  constructor({config, httpsAgent} = {}) {
+    this.client_id = config.client_id;
+    this.client_secret = config.client_secret;
+    this.token_endpoint = config.token_endpoint;
+    this.grant_type = config.grant_type;
+    this.scope = config.scope;
+    this.maxRetries = config.maxRetries;
+    this.cacheMax = config.cacheMax;
+    this.httpsAgent = httpsAgent;
+  }
+
+  /**
+   * Gets a new access token by first checking the database to see if an
+   * access token already exists for the client id; if not it creates an
+   * access token then stores it in the database.
+   *
+   * @returns {string} The access token.
+   */
+  async getAccessToken() {
+    const {
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries, cacheMax = 10
+    } = this;
+    const cache = new LruCache({
+      max: cacheMax,
+      // 2 weeks
+      maxAge: 2 * 7 * 24 * 60 * 60 * 1000
+    });
```

Comment:
So this cache needs to be created at the module level. Doing it here will just cause it to go out of scope and get garbage collected as soon as `getAccessToken()` returns.

Also, the config should come from the `bedrock.config` for this module. As for the defaults, we want to be able to store as many access tokens as a single service would be reasonably using -- without more knowledge about that I'd just go with `100`. As for the max age, this cache is only used to avoid hitting the authz server and the database too frequently -- it isn't related to how long the access tokens live. I'd set it to 5 minutes by default.

Reply:
Updated, thanks for the clarification.